platform-api/src/main/java/io/datacater/core/kubernetes/PythonRunnerPool.java

```diff
@@ -116,206 +94,101 @@ private String prependSlash(String path) {
-        List<Pod> pods =
-            kubernetesClient
-                .pods()
-                .inNamespace(DataCaterK8sConfig.NAMESPACE)
-                .withLabels(DataCaterK8sConfig.LABELS)
-                .list()
-                .getItems();
-        Deque<NamedPod> namedPods =
-            new ArrayDeque<NamedPod>(DataCaterK8sConfig.PYTHON_RUNNER_REPLICAS);
-        for (Pod pod : pods) {
-          String podName = pod.getMetadata().getName();
-          namedPods.add(new NamedPod(podName, pod));
-        }
-
-        return map.get(POOL_NAME)
-            .onItem()
-            .ifNull()
-            .continueWith(new RunnerPool(namedPods))
-            .invoke(pool -> map.put(POOL_NAME, pool))
-            .onItem()
-            .ifNotNull()
-            .invoke(pool -> map.replace(POOL_NAME, pool));
-      };
-
-  Function<Lock, Pod> lockTransformerUni =
-      lock -> {
-        Uni<AsyncMap<String, Pod>> pods = sharedData.getAsyncMap(POOL_NAME);
-        lock.release();
-        return pods.onItem().transform(locki -> new Pod()).await().indefinitely();
-      };
-
-  Function<AsyncMap<String, Deque<NamedPod>>, NamedPod> transformAsyncMapToPod =
-      map ->
-          map.get(POOL_NAME)
-              .onItem()
-              .transform(Deque::pop)
-              .onFailure()
-              .transform(
-                  Unchecked.function(
-                      throwable -> {
-                        throw new DatacaterException(throwable.getMessage());
-                      }))
-              .await()
-              .indefinitely();
-
-  UnaryOperator<Pod> annotatePodAsUsed =
-      pod -> {
-        Map<String, String> annotations = pod.getMetadata().getAnnotations();
-        String podName = pod.getMetadata().getName();
-
-        if (annotations.containsKey(IN_USE_LABEL_KEY)
-            && annotations.get(IN_USE_LABEL_KEY).equals(IN_USE_TRUE_LABEL_VALUE)) {
-          throw new DatacaterException(
-              String.format(
-                  "Pod with Name := %s is annotated as in use. Will not re-annotate.", podName));
-        }
-
-        return kubernetesClient
-            .pods()
-            .inNamespace(DataCaterK8sConfig.NAMESPACE)
-            .withName(podName)
-            .edit(
-                p ->
-                    new PodBuilder(p)
-                        .editOrNewMetadata()
-                        .addToAnnotations(IN_USE_TRUE_LABEL)
-                        .endMetadata()
-                        .build());
-      };
-
-  private Boolean isStatefulSetActive() {
-    int statefulSetCount =
+  private Deque<NamedPod> podsFromKubernetes() {
+    List<Pod> pods =
         kubernetesClient
-            .apps()
-            .statefulSets()
+            .pods()
             .inNamespace(DataCaterK8sConfig.NAMESPACE)
             .withLabels(DataCaterK8sConfig.LABELS)
+            .withField("status.phase", "Running")
             .list()
-            .getItems()
-            .size();
```

Comment:
Could we come into the scenario, where we de-queue a runner which then executes for a while, reset the queue, and the runner get re-added and used, even though it is still executing?

I don't see anything marking the runners is in an `executing`-state.

If it's relevant, we could open an improvement issue.

Reply:
We should test thoroughly as it might be already by way of ordering the `getMap` request by smallrye mutiny. Just in case I created a ticket https://github.com/DataCater/datacater/issues/27.